File: problems/0005-longest-palindromic-substring/solution.py
"""
Problem Number: 5
Problem Name: Longest Palindromic Substring
Difficulty: Medium
Topics: String, Dynamic Programming
Companies: Amazon, Google, Microsoft, Meta
"""

class Solution:
    def longestPalindrome(self, s: str) -> str:
        """
        Approach: Expand Around Center
        Time Complexity: O(n²)
        Space Complexity: O(1)
        """
        res = ""
        resLen = 0
        
        for i in range(len(s)):
            # Odd length
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > resLen:
                    res = s[l:r+1]
                    resLen = r - l + 1
                l -= 1
                r += 1
            
            # Even length
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > resLen:
                    res = s[l:r+1]
                    resLen = r - l + 1
                l -= 1
                r += 1
                
        return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.longestPalindrome("babad")) # "bab" or "aba"
    print(solution.longestPalindrome("cbbd"))  # "bb"

File: problems/0005-longest-palindromic-substring/solution.js
/**
 * @problemNumber: 5
 * @problemName: Longest Palindromic Substring
 * @difficulty: Medium
 * @topics: String, Dynamic Programming
 * @companies: Amazon, Google, Microsoft, Meta
 */

/**
 * Approach: Expand Around Center
 * Time Complexity: O(n²)
 * Space Complexity: O(1)
 * 
 * @param {string} s
 * @return {string}
 */
function longestPalindrome(s) {
    if (!s || s.length < 1) return "";

    let start = 0;
    let end = 0;

    for (let i = 0; i < s.length; i++) {
        const len1 = expandAroundCenter(s, i, i);
        const len2 = expandAroundCenter(s, i, i + 1);
        const len = Math.max(len1, len2);

        if (len > end - start) {
            start = i - Math.floor((len - 1) / 2);
            end = i + Math.floor(len / 2);
        }
    }

    return s.substring(start, end + 1);
}

function expandAroundCenter(s, left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        left--;
        right++;
    }
    return right - left - 1;
}

// Test cases
console.log(longestPalindrome("babad")); // "bab" or "aba"
console.log(longestPalindrome("cbbd"));  // "bb"

module.exports = longestPalindrome;
